nesta/packages/vectors/read.py

```diff
@@ -0,0 +1,109 @@
+        a_vector, = session.query(orm.vector).limit(1).one()
+        dim = len(a_vector) # "Width" of array
+    # Preallocate space
+    data = np.empty((count, dim), dtype=FLOAT_TYPE)
+    ids = np.empty((count, ), dtype=STR_TYPE)
+    return data, ids
+
+
+def read_data(data, ids, orm, id_field, database,
+              chunksize=10000, max_chunks=None,
+              section="mysqldb", db_env="MYSQLDB"):
+    """Read data into the data and id arrays, 
+    always starting from the last read chunk (e.g. if connection fails). 
+    Data is read using the last available id,
+    since filtering is much faster than offsetting, for large datasets.
+    """
+    id_field = getattr(orm, id_field)
+    fields = (id_field, orm.vector)
+    count, _ = data.shape
+    empty_ids = (ids != '')
+    offset = sum(empty_ids)  # resume if already started
+    # Calculate a filter statement, since these are faster than OFFSET
+    filter_ = None if offset == 0 else id_field > ids[offset-1]
+    # Prepare the engine for interacting with the DB
+    engine = get_mysql_engine(db_env, section, database)
+    # Set default values of {n,max}_chunks
+    n_chunks = -1 if max_chunks is None else 0
+    max_chunks = 0 if max_chunks is None else max_chunks
+    # Start or continue collecting filling the data and id arrays
+    while offset < count:
+        if n_chunks >= max_chunks:
+            # Note: this never happens if max_chunks is set to default
+            break
+        if offset % 10*chunksize == 0:
+            logging.info(f"Collecting row {offset+1} of {count}")
+        # Query the database and bundle the results into intermediate arrays
+        limit = chunksize if offset + chunksize < count else None
+        with db_session(engine) as session:
+            _ids, _data = query_and_bundle(session, fields, offset, limit, filter_)
+        # Update the preallocated arrays
+        ids[offset:offset+_ids.shape[0]] = _ids
+        data[offset:offset+_data.shape[0]] = _data
+        # Update the filter/offset criteria
+        filter_ = id_field > _ids[-1]
+        offset += chunksize
+        # Increment the number of chunks we've processed thus far
+        if max_chunks > 0:
+            # Note: this never happens if max_chunks is set to default
+            n_chunks += 1
+
+
+def download_vectors(orm, id_field, database, 
+                     chunksize=10000, max_chunks=None):
+    """Download vectors from the DB"""
+    data, ids = prefill_inputs(orm, database)  # Empty numpy arrays
+    while "reading data":
+        try:
+            # Start or continue reading
+            read_data(data=data, ids=ids, orm=orm, 
+                      id_field=id_field, database=database,
+                      chunksize=chunksize, 
+                      max_chunks=max_chunks)
+        # The following has only been found to happen if your
+        # connection drops slightly, which corrupts the JSON
```

Comment:
How does the corruption happen?

Reply:
A while back I gave a prototype of the above code to Juan and we both had issues with JSON corruption, which I found went away by using ethernet. Debugging with an `except`, I found that a single character is occasionally altered in the middle of the JSON leading to bad JSON parsing. I can only assume that somehow it's related to how the data is unpacked over WiFi... seems like a potential security flaw to me, but I'm not sure I want to dig that deep!